### PR TITLE
EventHub 2024-05-01-preview

### PR DESCRIPTION
**Pull request by schaudhari6254888**
# ARM (Control Plane) API Specification Update Pull Request 

> [!TIP]
> Overwhelmed by all this guidance? See the `Getting help` section at the bottom of this PR description.

> [!NOTE]
> As of January 2024 there is no PR assignee. This is expected. See https://aka.ms/azsdk/pr-arm-review.

## PR review workflow diagram

Please understand this diagram before proceeding. It explains how to get your PR approved & merged.

diagram

<details>
<summary>Click here to see the details of Step 1, Breaking Changes review</summary>

If you are in purview of Step 1 of the diagram, follow the [Breaking Changes review process].  
**IMPORTANT!** This applies even if you believe your PR was mislabeled, for any reason, including tool failure.
</details>

<details> 
<summary>Click here to see the details of Step 2, ARM review</summary>

See https://aka.ms/azsdk/pr-arm-review.
</details>

<details>
<summary> Click here to see the diagram footnotes</summary>

### Diagram footnotes

[1] See [ARM review queue] (for **PR merge** queues, see [2]).  
[2] [public repo merge queue], [private repo merge queue] (for **ARM review** queue, [1])  
The ARM reviewer on-call engineer visits the merge queue twice a day, so the approximate ETA for merges is 12 - 24 hours.

</details>

## Purpose of this PR

What's the purpose of this PR? Check the specific option that applies. This is **mandatory**!

  - [ ] New resource provider.
  - [ ] New API version for an existing resource provider. (If API spec is not defined in TypeSpec, the PR should have been generated using [OpenAPI Hub](https://aka.ms/openapihub)).
  - [ ] Update existing version for a new feature. (This is applicable only when you are revising a private preview API version.)
  - [ ] Update existing version to fix OpenAPI spec quality issues in S360.
  - [ ] Other, please clarify:
    - _edit this with your clarification_

## Due diligence checklist

To merge this PR, you **must** go through the following checklist and confirm you understood 
and followed the instructions by checking all the boxes:

- [ ] I confirm this PR is modifying Azure Resource Manager (ARM) related specifications, and not data plane related specifications.
- [ ] I have reviewed following [Resource Provider guidelines](https://aka.ms/rpguidelines), including
  [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and
  [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md) (estimated time: 4 hours).  
  I understand this is required before I can proceed to the diagram Step 2, "ARM API changes review", for this PR.

## Additional information

<details>
<summary> Viewing API changes</summary>

For convenient view of the API changes made by this PR, refer to the URLs provided in the table 
in the `Generated ApiView` comment added to this PR. You can use ApiView to show API versions diff. 

</details>
<details>
<summary>Suppressing failures</summary>

If one or multiple validation error/warning suppression(s) is detected in your PR, please follow the 
[suppressions guide](https://aka.ms/azsdk/pr-suppressions) to get approval.

</details>

## Getting help

- First, please carefully read through this PR description, from top to bottom. Please fill out the `Purpose of this PR` and `Due diligence checklist`.
- To understand what you must do next to merge this PR, see the `Next Steps to Merge` comment. It will appear within few minutes of submitting this PR and will continue to be up-to-date with current PR state.
- For guidance on fixing this PR CI check failures, see the hyperlinks provided in given failure 
  and https://aka.ms/ci-fix.
- For help with PR workflow diagram Step 2 (ARM review), see https://aka.ms/azsdk/pr-arm-review.
- If the PR CI checks appear to be stuck in `queued` state, please add a comment with contents `/azp run`.
  This should result in a new comment denoting a `PR validation pipeline` has started and the checks should be updated after few minutes.
- If the help provided by the previous points is not enough, post to https://aka.ms/azsdk/support/specreview-channel and link to this PR.

[ARM review queue]: https://aka.ms/azsdk/pr-arm-review#about-the-arm-review-queue
[public repo merge queue]: https://github.com/Azure/azure-rest-api-specs/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+draft%3Afalse+sort%3Acreated-asc
[private repo merge queue]: https://github.com/Azure/azure-rest-api-specs-pr/pulls?q=is%3Aopen+is%3Apr+label%3AMergeRequested+-label%3AApproved-OkToMerge+draft%3Afalse+sort%3Acreated-asc
[Breaking Changes review process]: https://eng.ms/docs/cloud-ai-platform/azure-core/azure-core-pm-and-design/trusted-platform-pm-karimb/service-lifecycle-and-actions-team/service-lifecycle-actions-team/apex/media/launchingproductbreakingchanges#breaking-change-process-1

### CODE REVIEW

**Comment by ms-henglu** [comment]
There are a bunch of breaking changes reported on this PR. Please get a breaking changes approval from Azure Breaking Changes Reviewers <azbreakchangereview@microsoft.com> and ARM can review the changes afterwards. Once you have the approval, please remove the "ARMChangesRequested" label from the PR to make the PR visible to the ARM reviewers.

File: specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/examples/NameSpaces/EHNamespaceFailover.json
Comment (mikeharder) [inline specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/examples/NameSpaces/EHNamespaceFailover.json:18]: I believe this header value is causing a validation failure from AutoRest:

```
    {
      code: 'MIN-LENGTH',
      params: [ 6, 10 ],
      message: 'String is too short (6 chars), minimum 10',
      path: '#/paths/~1subscriptions~1{subscriptionId}~1resourceGroups~1{resourceGroupName}~1providers~1Microsoft.EventHub~1namespaces~1{namespaceName}~1failover/post/x-ms-examples/NameSpaceCreate/responses/202/headers/location',
      schemaId: undefined,
      description: 'An example of the location header sent by the server, if applicable in the example.'
    },
```

I believe the value is supposed to be a URL like this:

https://github.com/Azure/azure-rest-api-specs/blob/3e81da30a1358b6ffcbc7ff222aae5f483971d32/specification/advisor/resource-manager/Microsoft.Advisor/stable/2023-01-01/examples/GenerateRecommendations.json#L9

Comment (schaudhari6254888) [reply specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/examples/NameSpaces/EHNamespaceFailover.json:18]: Hey @mikeharder, we have done the same thing in the past API-2023-01-01-preview version.

https://github.com/Azure/azure-rest-api-specs/blob/3e81da30a1358b6ffcbc7ff222aae5f483971d32/specification/eventhub/resource-manager/Microsoft.EventHub/preview/2023-01-01-preview/examples/NameSpaces/EHNamespaceFailover.json#L18

Comment (mikeharder) [reply specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/examples/NameSpaces/EHNamespaceFailover.json:18]: > Hey @mikeharder, we have done the same thing in the past API-2023-01-01-preview version.
> 
> https://github.com/Azure/azure-rest-api-specs/blob/3e81da30a1358b6ffcbc7ff222aae5f483971d32/specification/eventhub/resource-manager/Microsoft.EventHub/preview/2023-01-01-preview/examples/NameSpaces/EHNamespaceFailover.json#L18

I get the same error on this version of the spec:

```
$ npx autorest --v2 --input-file=specification/eventhub/resource-manager/Microsoft.EventHub/preview/2023-01-01-preview/namespaces-preview.json
```

If your question is "why didn't we see this error earlier", it's because some of our checks had bugs that would hide errors like this.

The correct path forward is to fix all the wrong examples.

**Comment by schaudhari6254888** [comment]
> There are a bunch of breaking changes reported on this PR. Please get a breaking changes approval from Azure Breaking Changes Reviewers [azbreakchangereview@microsoft.com](mailto:azbreakchangereview@microsoft.com) and ARM can review the changes afterwards. Once you have the approval, please remove the "ARMChangesRequested" label from the PR to make the PR visible to the ARM reviewers.

Hey @ms-henglu , Now there is no breaking on the PR
Can you pls review the PR now.

**Comment by schaudhari6254888** [comment]
> /azp run

Hey @mikeharder , can we suppressed the Swagger LinDiff, Avacado Pipeline error here?
Also Where can I get the `sdk-for-js` Approval

**Comment by rkmanda** [comment]
> > /azp run
> 
> Hey @mikeharder , can we suppressed the Swagger LinDiff, Avacado Pipeline error here? Also Where can I get the `sdk-for-js` Approval

Whats the reason for not being able to fix the lintdiff errors? There are a ton of violations that cannot be approved. https://github.com/Azure/azure-rest-api-specs/pull/28450/checks?check_run_id=23489128434

**Comment by schaudhari6254888** [comment]
> > > /azp run
> > 
> > 
> > Hey @mikeharder , can we suppressed the Swagger LinDiff, Avacado Pipeline error here? Also Where can I get the `sdk-for-js` Approval
> 
> Whats the reason for not being able to fix the lintdiff errors? There are a ton of violations that cannot be approved. https://github.com/Azure/azure-rest-api-specs/pull/28450/checks?check_run_id=23489128434

Hey @rkmanda, I have copied from 2023-01-01-preview and renamed it to 2024-01-01-preview with few more addition. The point is the previous version is also we got approval. this LintDiff error are coming from the files that are exact same as 2023-01-01-preview.

**Comment by schaudhari6254888** [comment]
> > > > /azp run
> > > 
> > > 
> > > Hey @mikeharder , can we suppressed the Swagger LinDiff, Avacado Pipeline error here? Also Where can I get the `sdk-for-js` Approval
> > 
> > 
> > Whats the reason for not being able to fix the lintdiff errors? There are a ton of violations that cannot be approved. https://github.com/Azure/azure-rest-api-specs/pull/28450/checks?check_run_id=23489128434
> 
> Hey @rkmanda, I have copied from 2023-01-01-preview and renamed it to 2024-01-01-preview with few more addition. The point is the previous version is also we got approval. this LintDiff error are coming from the files that are exact same as 2023-01-01-preview.

For the reference I am adding the Old PR
https://github.com/Azure/azure-rest-api-specs/pull/26672
This is the PR where we got LintDiff error approved.
If we fix something here many SDK which will generate from this Swagger will cause of backward compatibility of SDK.
CC - @mikeharder

File: specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/namespaces.json
Comment (gary-x-li) [inline specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/namespaces.json:182]: Can you also add `x-ms-long-running-operation-options` to indicate which header to use? This applies to all other occurrences.

File: specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/namespaces.json
Comment (gary-x-li) [inline specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/namespaces.json:268]: Also add `x-ms-long-running-operation` and `x-ms-long-running-operation-options`?

Comment (schaudhari6254888) [reply specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/namespaces.json:274]: Added

File: specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/namespaces.json
Comment (schaudhari6254888) [reply specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/namespaces.json]: Added

File: specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/namespaces.json
Comment (gary-x-li) [reply specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/namespaces.json:274]: They don't seem to be added under `patch`.

Comment (schaudhari6254888) [reply specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/namespaces.json:274]: Hey @gary-x-li  ,this change is causing too much CI checks failed because previous swagger of a namespace doesn't have this  `x-ms-long-running-operation` property
https://github.com/schaudhari6254888/azure-rest-api-specs/blob/716b241d1cc21b7830b7d5bfaa605d85501668f0/specification/eventhub/resource-manager/Microsoft.EventHub/stable/2024-01-01/namespaces.json#L265

Can you pls approve is this change is not very concerning, this PR is blocking future SDK release.

**Review by gary-x-li — APPROVED** [review]

**Comment by kazrael2119** [comment]
@schaudhari6254888 we met a duplicate model issue when preparing JS sdk for eventhub 2024-05-01-preview from
https://github.com/Azure/azure-rest-api-specs/blob/main/specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/namespaces.json#L385
this references the ErrorResponse model from **common-type** which is different with others referenced from **Microsoft.EventHub/common/**

**Comment by wanlwanl** [comment]
FYI. https://dev.azure.com/azure-sdk/internal/_build/results?buildId=3688010&view=logs&j=47092895-c7c8-567a-c7a6-8c88f3ec178e&t=d293a8e8-c6af-5091-cdf7-4c66482dd6f2&l=504

**Comment by schaudhari6254888** [comment]
> @schaudhari6254888 we met a duplicate model issue when preparing JS sdk for eventhub 2024-05-01-preview. From https://github.com/Azure/azure-rest-api-specs/blob/main/specification/eventhub/resource-manager/Microsoft.EventHub/preview/2024-05-01-preview/namespaces.json#L385, here references the ErrorResponse model from **common-type** which is different with others referenced from **Microsoft.EventHub/common/**

Hey @kazrael2119, I got the issue. I have sent out the fix.
Fix- https://github.com/Azure/azure-rest-api-specs/pull/32508

**Pull request merged**
